src/parser.cpp

```diff
@@ -400,6 +392,107 @@ std::unique_ptr<Expr> Parser::grouping()
     return expr;
 }
 
+std::unique_ptr<Stmt> Parser::func()
+{
+    if (this->advance().token_type != Token::Type::FN)
+    {
+        throw BirdException("expected fn keyword");
+    }
+
+    auto fn_identifier = this->advance();
+
+    if (fn_identifier.token_type != Token::Type::IDENTIFIER)
+    {
+        this->user_error_tracker->expected("identifier", "after keyword", this->peek_previous());
+        this->synchronize();
+        throw UserException();
+    }
+    if (this->advance().token_type != Token::Type::LPAREN)
+    {
+        this->user_error_tracker->expected("(", "after identifier", this->peek_previous());
+        this->synchronize();
+        throw UserException();
+    }
+
+    auto fn_params = this->fn_params();
+
+    if (this->advance().token_type != Token::Type::RPAREN)
+    {
+        this->user_error_tracker->expected(")", "after function parameter list", this->peek_previous());
+        this->synchronize();
+        throw UserException();
+    }
+
+    auto fn_return_type = this->fn_return_type();
+
+    auto block = this->block();
+
+    return std::make_unique<Func>(Func(fn_identifier, fn_return_type, fn_params, std::move(block)));
+}
+
+std::vector<std::pair<Token, Token>> Parser::fn_params()
+{
+    if (this->peek().token_type == Token::Type::RPAREN)
+    {
+        return {};
+    }
+
+    std::vector<std::pair<Token, Token>> params;
+
+    while (true)
+    {
+        params.push_back(this->param_decl());
+
+        if (this->peek().token_type == Token::Type::RPAREN)
+        {
+            return params;
+        }
+        else if (this->advance().token_type != Token::Type::COMMA)
+        {
+            this->user_error_tracker->expected(",", "after function parameter", this->peek_previous());
+            this->synchronize();
+            throw UserException();
+        }
+    }
+}
+
+std::pair<Token, Token> Parser::param_decl()
+{
+    auto identifier = this->advance(); // TODO: check that this is an identifier
+
+    if (this->advance().token_type != Token::Type::COLON)
+    {
+        this->user_error_tracker->expected(":", "after identifier in assignment", this->peek_previous());
```

Comment:
Can you change this error message to something more appropriate? It isn't technically an assignment, so maybe say "parameter list"?